ConditionalCompilation.md

```diff
@@ -1,28 +1,139 @@
 # Conditional Compilation
 
-(TBD)
-
-This section will discuss extensions to WGSL
-to support conditional compilation, where the conditions
-are set by the host code runtime or build time.
-
-* `#if #else #endif` (or perhaps `#ifdef`) in the style of a C preprocessor 
-  is the approach that comes first to mind.
-* conditional compilation runs after simple templating but before import processing..
-  * (note that we may want to disable code that may not be correct so parsing first is iffy)
-* best to define default values in the wgsl for all conditionals
-  * IDE tools can't assume any definitions that aren't in the wgsl.
-* alternately, consider some kind of [cfg] on elements instead? 
-  * More structured: ties to source elements not lines
-    * Perhaps some optimization potential..
-      * For example could parse/lower certain functions ahead of 
-        time and easily remove them from the IR vs #if #else which operates on a source level.
-    * Is it an wgsl attribute? syntax makes sense but wgsl 
-* what are some examples of things to conditionally compile?
-  * function variants e.g. select which function variant to use based on condition
-  * struct fields e.g. include fields only if condition is set
-  * import statements? e.g. import util/fast vs util/regular based on condition
-* how do we support build time vs runtime linking?
-  * e.g. some projects will want prebuilt versions for various mobile configurations.
-    * ref [unity shader conditionals](https://docs.unity3d.com/Manual/shader-conditionals.html)
-  * should the wesl code distinguish conditions that are settable at compile time from those that are settable at runtime?
+## Overview
+
+> This section is non-normative
+
+Conditional compilation is a mechanism to modify the source code based on parameters passed to the compiler.
+This specificaton extends the [*attribute* syntax](https://www.w3.org/TR/WGSL/#attributes) with three new attributes: `@if`, `@elif`, `@else`.
+These attributes indicate that the syntax node they decorate can be removed by the compiler based on feature flags.
+
+> Note: This implementation is similar to the `#[cfg(feature = "")]` syntax in rust.
+
+### Usage Example
+
+```rs
+// global variables and bindings...
+@if(textured)
+@group(0) @binding(0) var my_textuer: texture_2d<f32>;
+
+// structs declarations and struct members...
+struct Ray {
+  position: vec4f,
+  direction: vec4f,
+  @if(debug_mode && raytracing_enabled)
+  ray_steps: u32,
+}
+
+// function declarations, parameters and statements...
+fn main() -> vec4 {
+  @if(legacy_implementation || (is_web_version && xyz_not_supported))
+  let result = legacy_impl();
+  @else
+  let result = modern_impl();
+}
+```
+
+## Definitions
+
+ * **Compile-time expression**: A *compile-time expression* is evaluated by the compiler before the rest of the program. It lives in the *compile-time scope*.
```

Comment:
I was a little confused by the phrase 'compile-time' (as you saw in my earlier comments). I hear 'compile time' and think that must be running at build time, and so settings will be fixed at runtime (e.g. the gpu capabilities). Of course here compilation may happen at runtime, instead of at build time. 

Lest others be confused, I wonder if we might use a different word than compile-time. transpile-time, assembly-time, perhaps? Or if there's no good alternative phrase, then I'd suggest an explanation right up front that compilation can happen at runtime.

Reply:
"Translation time"? And we just define that word up front.

Reply:
+1 for translation time!

(but agree that explaining up front is sufficient if no alternate phrase appeals)

Reply:
OK for me for translation time. could add that to the glossary too. Also, I never know if we should be using linker/compiler/transpiler/preprocessor... I think it would be favorable to decide that, as the term will appear in the spec. My preference is compiler.

Reply:
Renamed compilation to translation and renamed compiler to WESL translator.
Not a big fan of WESL translator, but I chose it for consistency.

Open to bikeshedding :)